_posts/2024-06-10-what-is-a-list-anyway.md

```diff
@@ -0,0 +1,252 @@
+"returns" a new list. What we are declaring is a relation from one list to another and we can do it
+the other way as well or use it to check if the relation holds for the given parameters:
+
+```prolog
+?- maplist(succ, MinusOnes, [1,2,3,4]).
+MinusOnes = [0, 1, 2, 3].
+
+?- maplist(succ, [1,2], [2,3]).
+true.
+```
+
+Next, let's define relational versions of the usual list functions: drop (remove N items from front)
+and take (take only the N first items of the list).
+
+```prolog
+drop(0, Lst, Lst).
+drop(N, [_|Rest], Out) :-
+  N > 0,
+  N1 is N - 1,
+  drop(N1, Rest, Out).
+```
+
+Above you can see we define two clauses. The first states that when dropping 0 elements
+from a list, the output list is the list itself. The second states that for a positive
+number N, the first element of the input list is discarded, and then a recursive clause
+is added for N-1.
+
+Similarly, we can define take as:
+```prolog
+take(0, _, []).
+take(N, [I|Items], [I|Taken]) :-
+  N > 0,
+  N1 is N - 1,
+  take(N1, Items, Taken).
+```
+
+Again we have a recursive definition with a base case of taking zero items, but here
+the output of taking zero items from anything is the empty list.
+
+We can now use both of these definitions:
+```prolog
+?- drop(2, [1,2,3,4], Dropped).
+Dropped = [3, 4].
+
+?- take(3, [42, 1, 2, 3, 4], Taken).
+Taken = [42, 1, 2].
+```
+
+As these are relations, can we run them the other way? Yes, somewhat.
+
+```prolog
+?- between(1,3,N), drop(N, Input, [1,2,3]).
+N = 1,
+Input = [_, 1, 2, 3];
+
+N = 2,
+Input = [_, _, 1, 2, 3];
+
+N = 3,
+Input = [_, _, _, 1, 2, 3].
+```
+
+We can state that the result of dropping N items is the list `[1,2,3]` and Prolog
+will give us the possible input lists, but all the items before will be
+placeholders that can refer to anything.
+
+But we can further constrain the list and state that all numbers are between 1 and 3 to
+get all possible permutations:
+
+```prolog
+?- between(1,3,N), drop(N, Input, [1,2,3]), maplist(between(1,3), Input).
+N = 1,
+Input = [1, 1, 2, 3];
+
+N = 1,
+Input = [2, 1, 2, 3];
+
+N = 1,
+Input = [3, 1, 2, 3];
+
+N = 2,
+Input = [1, 1, 1, 2, 3];
+
+N = 2,
+Input = [1, 2, 1, 2, 3];
+
+% ...many results omitted...
+
+N = 3,
+Input = [3, 3, 3, 1, 2, 3].
+```
+
+## Enter append
+
+Of course, a language must also have a way to concatenate lists. For this, we have `append` which comes
+in two forms: append two lists to get a third one and append a list of lists to get all of them concatenated.
+
+```prolog
+?- append([h,e], [l,l,o], Out).
+Out = [h, e, l, l, o].
+
+?- append([[1,2], [3,4], [5,6]], All).
+All = [1, 2, 3, 4, 5, 6].
+```
+
+Pretty simple... a utility found in most languages.
+What if I told you that the above utilities of `drop` and `take` aren't needed
+because we have `append`? How can appending lists be used to split them apart, I hear you ask.
+Remember that we are describing a relation between the parameters, so can do drop and take like:
```

Comment:
> so can do drop and take like

This looks like it's missing something?

-> so *we* can do drop and take like ?